File: examples/core/page-rank/src/main/scala/my/spark/core/PageRank.scala
package my.spark.core

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._

object PageRank {
  def main(args: Array[String]) {
    val sc = new SparkContext()
    
    val data = List("page1 page3","page2 page1","page4 page1","page3 page1","page4 page2","page3 page4")

    val links = sc.parallelize(data).map(_.split(" ")).map(pages => (pages(0),pages(1))).distinct().groupByKey().cache()
    println("\n(Source, (Target List))")
    links.collect().foreach { case (page,list) => println("(" + page + ",(" + list.mkString(", ") + "))") }

    var ranks = links.map(pair => (pair._1, 1.0))
    println("\n(Page, InitialRank)")
    ranks.collect().foreach(println)

    for (x <- 1 to 10) {
      var contribs = links.join(ranks).flatMap(pair => computeContributions(pair._2._1, pair._2._2)) 
      ranks = contribs.reduceByKey(_+_).map { case (page,rank) => (page, rank * 0.85 + 0.15) }
      println("\nIteration " + x)
      for (rankpair <- ranks.collect()) println(rankpair)    
    }
    println()
    sc.stop()
  }

  def computeContributions(neighbors: Iterable[String], rank: Double): Iterable[(String,Double)] = {
    for (neighbor <- neighbors) yield(neighbor, rank/neighbors.size)
  }
}

File: examples/core/page-rank/src/main/python/PageRank.py
import sys
from pyspark import SparkContext

if __name__ == "__main__":

  def computeContribs(neighbors, rank):
    for neighbor in neighbors: yield(neighbor, rank/len(neighbors))

  sc = SparkContext()

  data = ["page1 page3","page2 page1","page4 page1","page3 page1","page4 page2","page3 page4"]

  links = sc.parallelize(data).map(lambda line: line.split()).map(lambda pages: (pages[0],pages[1])).distinct().groupByKey().cache()
  print("\n(Source, (Target List))")
  for link in links.collect(): print '({0}, ({1}))'.format(link[0], ', '.join(link[1]))

  ranks = links.map(lambda (page,neighbors): (page,1.0))
  print("\n(Page, InitialRank)")
  for rank in ranks.collect(): print '({0}, {1})'.format(rank[0], rank[1])

  for x in xrange(10):
    contribs = links.join(ranks).flatMap(lambda (page,(neighbors,rank)): computeContribs(neighbors,rank))
    ranks = contribs.reduceByKey(lambda v1,v2: v1+v2).map(lambda (page,contrib): (page,contrib * 0.85 + 0.15))
    print "\nIteration ", x+1
    for pair in ranks.take(10): print '({0}, {1})'.format(pair[0], pair[1])
  print
